### PR TITLE
Store a index per dep node kind

### PR DESCRIPTION
**Pull request by Zoxc**
This stores an index map per dep node kind instead of using a single index map. This avoids storing `DepKind` in the key for greater cache efficiency.

<table><tr><td rowspan="2">Benchmark</td><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th></tr><tr><td align="right">Time</td><td align="right">Time</td><td align="right">%</th><td align="right">Memory</td><td align="right">Memory</td><td align="right">%</th></tr><tr><td>🟣 <b>clap</b>:check:unchanged</td><td align="right">0.4376s</td><td align="right">0.4334s</td><td align="right"> -0.95%</td><td align="right">92.53 MiB</td><td align="right">92.39 MiB</td><td align="right"> -0.15%</td></tr><tr><td>🟣 <b>hyper</b>:check:unchanged</td><td align="right">0.1442s</td><td align="right">0.1446s</td><td align="right"> 0.29%</td><td align="right">47.48 MiB</td><td align="right">47.63 MiB</td><td align="right"> 0.31%</td></tr><tr><td>🟣 <b>regex</b>:check:unchanged</td><td align="right">0.3296s</td><td align="right">0.3261s</td><td align="right">💚  -1.06%</td><td align="right">72.37 MiB</td><td align="right">71.20 MiB</td><td align="right">💚  -1.62%</td></tr><tr><td>🟣 <b>syn</b>:check:unchanged</td><td align="right">0.6149s</td><td align="right">0.6014s</td><td align="right">💚  -2.19%</td><td align="right">105.74 MiB</td><td align="right">101.69 MiB</td><td align="right">💚  -3.83%</td></tr><tr><td>🟣 <b>syntex_syntax</b>:check:unchanged</td><td align="right">1.4882s</td><td align="right">1.4560s</td><td align="right">💚  -2.16%</td><td align="right">209.75 MiB</td><td align="right">201.46 MiB</td><td align="right">💚  -3.95%</td></tr><tr><td>Total</td><td align="right">3.0144s</td><td align="right">2.9616s</td><td align="right">💚  -1.75%</td><td align="right">527.87 MiB</td><td align="right">514.37 MiB</td><td align="right">💚  -2.56%</td></tr><tr><td>Summary</td><td align="right">1.0000s</td><td align="right">0.9879s</td><td align="right">💚  -1.21%</td><td align="right">1 byte</td><td align="right">0.98 bytes</td><td align="right">💚  -1.85%</td></tr></table>

### CODE REVIEW

File: compiler/rustc_query_system/src/dep_graph/serialized.rs [hunk 1]
```diff
@@ -81,8 +82,9 @@ pub struct SerializedDepGraph<K: DepKind> {
     /// A flattened list of all edge targets in the graph, stored in the same
     /// varint encoding that we use on disk. Edge sources are implicit in edge_list_indices.
     edge_list_data: Vec<u8>,
-    /// Reciprocal map to `nodes`.
-    index: FxHashMap<DepNode<K>, SerializedDepNodeIndex>,
+    /// Stores a map from fingerprints to nodes per dep node kind.
+    /// This is the reciprocal of `nodes`.
+    index: Vec<FxHashMap<PackedFingerprint, SerializedDepNodeIndex>>,
```
Comment (cjgillot) [inline compiler/rustc_query_system/src/dep_graph/serialized.rs:87]: What about a `UnHashMap`?

Comment (Zoxc) [reply compiler/rustc_query_system/src/dep_graph/serialized.rs]: That probably doesn't hurt.

**Comment by Zoxc** [comment]
I included a commit where dep kind counts are also encoded so the index hash maps can get the right initial capacity. That's an additional improvement:

<table><tr><td rowspan="2">Benchmark</td><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th></tr><tr><td align="right">Time</td><td align="right">Time</td><td align="right">%</th><td align="right">Memory</td><td align="right">Memory</td><td align="right">%</th></tr><tr><td>🟣 <b>clap</b>:check:unchanged</td><td align="right">0.4117s</td><td align="right">0.4096s</td><td align="right"> -0.49%</td><td align="right">93.09 MiB</td><td align="right">92.73 MiB</td><td align="right"> -0.39%</td></tr><tr><td>🟣 <b>hyper</b>:check:unchanged</td><td align="right">0.1233s</td><td align="right">0.1226s</td><td align="right"> -0.57%</td><td align="right">47.85 MiB</td><td align="right">47.81 MiB</td><td align="right"> -0.09%</td></tr><tr><td>🟣 <b>regex</b>:check:unchanged</td><td align="right">0.3025s</td><td align="right">0.3009s</td><td align="right"> -0.54%</td><td align="right">71.39 MiB</td><td align="right">71.05 MiB</td><td align="right"> -0.49%</td></tr><tr><td>🟣 <b>syn</b>:check:unchanged</td><td align="right">0.5760s</td><td align="right">0.5734s</td><td align="right"> -0.44%</td><td align="right">101.72 MiB</td><td align="right">101.55 MiB</td><td align="right"> -0.17%</td></tr><tr><td>🟣 <b>syntex_syntax</b>:check:unchanged</td><td align="right">1.4258s</td><td align="right">1.4197s</td><td align="right"> -0.43%</td><td align="right">202.54 MiB</td><td align="right">202.23 MiB</td><td align="right"> -0.15%</td></tr><tr><td>Total</td><td align="right">2.8392s</td><td align="right">2.8262s</td><td align="right"> -0.46%</td><td align="right">516.59 MiB</td><td align="right">515.37 MiB</td><td align="right"> -0.24%</td></tr><tr><td>Summary</td><td align="right">1.0000s</td><td align="right">0.9951s</td><td align="right"> -0.49%</td><td align="right">1 byte</td><td align="right">1.00 bytes</td><td align="right"> -0.26%</td></tr></table>

**Comment by cjgillot** [comment]
@rust-timer build df14cfe1d6418344f82966df5d7589e1e8682dda

**Comment by cjgillot** [comment]
Good max-rss and cycles gains.
@bors r+

**Comment by Zoxc** [comment]
That looks like a failure to retry.

**Comment by Kobzol** [comment]
The previous run didn't have any instruction count changes, but anyway the Max-RSS and especially cycle wins outweigh the few icount regressions.

@rustbot label: +perf-regression-triaged

**Pull request merged**